File: src/net/dinkla/raytracer/math/Vector2DF.java
package net.dinkla.raytracer.math;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 10.04.2010
 * Time: 15:17:35
 * To change this template use File | Settings | File Templates.
 */
public class Vector2DF extends Element2D {

    public Vector2DF(final float x, final float y) {
        super(x, y);
    }

    public Vector2DF plus(final Vector2DF v) {
        return new Vector2DF(x + v.x, y + v.y);
    }

    public Vector2DF minus(final Vector2DF v) {
        return new Vector2DF(x - v.x, y - v.y);
    }

    public Vector2DF mult(final float s) {
        return new Vector2DF(s * x, s * y);
    }

    public float dot(final Vector2DF v)  {
        return x * v.x + y * v.y;
    }

    public float dot(final Normal v)  {
        return x * v.x + y * v.y;
    }

    public Vector2DF normalize() {
        final float l = length();
        return new Vector2DF(x / l, y / l);
    }

}

File: src/net/dinkla/raytracer/math/Vector2DF.scala
package net.dinkla.raytracer.math

/**
  * Created by Dinkla on 10.07.2015.
  */
final class Vector2DF(x: Float, y: Float) extends Vector2D[Float](x, y) {

   override def length: Float = math.sqrt(sqrLength).toFloat

   override def unary_-(): Vector2D[Float] = new Vector2DF(-x, -y)

   override def plus(v: Vector2D[Float]): Vector2D[Float] = new Vector2DF(x + v.x, y + v.y)

   override def dot(v: Vector2D[Float]): Float = x * v.x + y * v.y

   override def normalize: Vector2D[Float] = {
     val l = length
     new Vector2DF(x / l, y / l)
   }

   override def toPoint: Point2D[Float] = new Point2DF(x, y)

 }
